Fix accidental deletion of sinterstore command when we meet wrong type error.

enjoy-binbin opened the pull request:
In `sunion` command, if we type any keys wrong(like type error), there will be an error
```
127.0.0.1:6381> flushdb
OK
127.0.0.1:6381> sadd c 1 2 3 
(integer) 3
127.0.0.1:6381> smembers c
1) "1"
2) "2"
3) "3"
127.0.0.1:6381> sunionstore c a b
(integer) 0
127.0.0.1:6381> smembers c
(empty array)
127.0.0.1:6381> 
127.0.0.1:6381> sadd c 1 2 3 
(integer) 3
127.0.0.1:6381> set a 123  // `a` is a string
OK
127.0.0.1:6381> sunion a b
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6381> sunion b a
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6381> sunionstore c a b
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6381> sunionstore c b a
(error) WRONGTYPE Operation against a key holding the wrong kind of value
```

In sinter command, it will be different. In sinter command, i think the result will be fine(although unlike sunion)
But in sinterstore command, the different will cause totally different about the `dstkey`.
```
127.0.0.1:6381> flushdb
OK
127.0.0.1:6381> sinter a b
(empty array)
127.0.0.1:6381> sinter b a
(empty array)
127.0.0.1:6381> set a 123
OK
127.0.0.1:6381> sinter a b
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6381> sinter b a
(empty array)
127.0.0.1:6381> sadd c 1 2 3
(integer) 3
// in here, there will be an type error, and `c(dstkey)` is exist
127.0.0.1:6381> sinterstore c a b
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6381> exists c
(integer) 1
// in here, we will delete the c(dstkey), although `a` is a string
127.0.0.1:6381> sinterstore c b a
(integer) 0
127.0.0.1:6381> exists c
(integer) 0
```

So i update `sinterGenericCommand`, and make them behave uniformly. After output:
```
127.0.0.1:6379> flushdb
OK
127.0.0.1:6379> sinter a b
(empty array)
127.0.0.1:6379> sinter b a
(empty array)
127.0.0.1:6379> set a 123
OK
127.0.0.1:6379> sinter a b
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6379> sinter b a
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6379> 
127.0.0.1:6379> sadd c 1 2 3
(integer) 3
127.0.0.1:6379> sinterstore c a b
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6379> exists c
(integer) 1
127.0.0.1:6379> sinterstore c b a
(error) WRONGTYPE Operation against a key holding the wrong kind of value
127.0.0.1:6379> exists c
(integer) 1
```

Also add more tests for some commands.
Mainly focus on 
- `wrong type error`: 
	expand test case (base on sinter bug) in non-store variant
	add tests for store variant (although it exists in non-store variant, i think it would be better to have same tests)
- the dstkey result when we meet `non-exist key (empty set)` in *store

sdiff:
- improve test case about wrong type error (the one we found in sinter, although it is safe in sdiff)
- add test about using non-exist key (treat it like an empty set)
sdiffstore:
- according to sdiff test case, also add some tests about `wrong type error` and `non-exist key`
- the different is that in sdiffstore, we will consider the `dstkey` result

sunion/sunionstore add more tests (same as above)

sinter/sinterstore also same as above ...

oranagra commented:
@enjoy-binbin good catch, thank you.
can you please add tests to cover these cases.

@redis/core-team this bug-fix is in some way a breaking change, please approve (for 7.0)

enjoy-binbin commented:
@oranagra Sure. I will do that when i have times. I also make pr about other commands. https://github.com/redis/redis/pull/9018 https://github.com/redis/redis/pull/9015  (not urgent, small fix. Hope you can take a look when you have time, would love to hear your thoughts)

madolson reviewed (dismissed):
LGTM

oranagra reviewed (commented):
@enjoy-binbin that's a lot of new test code (i didn't expect that).
are you sure many of these are not already covered?
can you please make a list (or a table) of what you added, so it'll be easier to review.
i.e. if some tests existed for the non-store variant and you added them for the store variant, or some existed for UNION and you added them for INTER, and so on..

also, there there are many multi-key commands involved with this test, maybe we should hold merging it to after #9033 so that we can make sure proper tagging are added to support external sharded test targets.

enjoy-binbin commented:
@oranagra Yes, many of these are not covered in some tests case. 
My mistake, i know it is quite difficult to review the test code. 
These codes are very similar. I added some notes in the top comment.
(But I'm not sure I described it clearly... sadly. It was a little bit hard for me to describe it) 

The reason i added extra tests for sdiff/sunion is that:
I think they are very similar in the first half of the code processing
Although there is no problem with the original code and has nothing to do with this fix
I think it might be better to add these test cases (in future?) Although it may be a bit long-winded

As for 9033. I am not very familiar with it. Will follow your suggestion
(ps: Sorry for the delay (saw it a long time age). It took me some time to block the language. Also busy with others things)
